File: source/GyeongHo/boj/10971.cpp
#include <iostream>
using namespace std;

int W[10][10];
bool visit[10];
int N, MIN = 987654321;

void costLow(int start, int loc, int sum, int cnt)
{
	if(cnt == N && start == loc) {
		if(MIN > sum) MIN = sum;
		return;
	}

	for(int i = 0; i < N; ++i) {
		if(W[loc][i] == 0) continue;

		if(visit[loc]) continue;
		visit[loc] = true;
		sum += W[loc][i];

		if(sum <= MIN) costLow(start, i, sum, cnt + 1);

		visit[loc] = false;
		sum -= W[loc][i];
	}
}

int main()
{
	cin >> N;
	for(int i = 0; i < N; ++i)
		for(int j = 0; j < N; ++j)
			cin >> W[i][j];

	for(int i = 0; i < N; ++i)
		costLow(i, 0, 0, 0);

	cout << MIN;
	return 0;
}

File: source/GyeongHo/boj/10971.cs
using System;
using System.Linq;

class TSP2
{
	static int N, MIN = 987654321;
	static bool[] visit = new bool[10];
	static int[,] W = new int[10,10];
	
	static void fullSearch(int start, int loc, int sum, int cnt)
	{
		if(cnt == N && start == loc) {
			if(sum < MIN) MIN = sum;
			return;
		}

		for(int i = 0; i < N; ++i) {
			if(visit[i] || W[loc, i] == 0) continue;
			visit[i] = true;
			sum += W[loc, i];
			if(sum < MIN) fullSearch(start, i, sum, cnt + 1);
			visit[i] = false;
			sum -= W[loc, i];
		}
	}

	static int Main()
	{
		N = Convert.ToInt32(Console.ReadLine());

		for(int i = 0; i < N; ++i) {
			var tmp = Console.ReadLine().Split(' ').Select(int.Parse).ToList();
			for(int j = 0; j < N; ++j)
				W[i, j] = tmp[j];
		}
		
		for(int i = 0; i < N; ++i)
			fullSearch(i, i, 0, 0);

		Console.Write(MIN);
		return 0;
	}
}
